### PR TITLE
NESpaceGrid rebased and squashed.

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

NESpaceGrid class refactored/ported from QMCHamiltonian/SpaceGrid with new input scheme and a decent start at test coverage.  Last component needed before the ported EnergyDensityEstimator can be PR'd.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
OSX, x86 v100

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
Test this please

File: src/Estimators/InputSection.h [hunk 1]
```diff
@@ -224,7 +224,7 @@ class InputSection
                                                 const std::string& name,
                                                 std::istringstream& svalue)
   {
-    throw std::runtime_error("derived class must provide handleCustom method if custom parameters are used");
```
Comment (ye-luo) [inline src/Estimators/InputSection.h:227]: Could you move these function definitions to the cpp file?

File: src/Estimators/InputSection.cpp [hunk 2]
```diff
@@ -74,22 +74,27 @@ void InputSection::readXML(xmlNodePtr cur)
...
-  if (!(checkSectionName(section_ename) || checkSectionName(section_method) ||
-        checkSectionName(section_type) || checkSectionName(section_name_actual)))
-    throw UniformCommunicateError("Input is invalid  " + lcase_section_name + " does not match input node!");
+  if (!(checkSectionName(section_ename) || checkSectionName(section_method) || checkSectionName(section_type) ||
+        checkSectionName(section_name_actual)))
+  {
+    std::stringstream error;
+    error << "Input is invalid \"" << lcase_section_name << "\" does not match a defined input node!";
+    throw UniformCommunicateError(error.str());
+  }
```
Comment (ye-luo) [inline src/Estimators/InputSection.cpp:97]: Are the changes in this file all just formatting?

**Review by ye-luo — APPROVED** [review]
I'm still having concerns on the use of `noreturn`. The standard says it is UB if the function returns. In a derived class, `noreturn` is inherited and gonna affect the override function. The standard is not very clear on inheritance.

Assuming we are testing the code properly. There is no need to resolve the issue in this PR.

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**